src/realm/object-store/sync/sync_session.cpp

```diff
@@ -1362,16 +1378,26 @@ void SyncSession::update_subscription_store(bool flx_sync_requested)
 void SyncSession::create_subscription_store()
 {
     REALM_ASSERT(!m_flx_subscription_store);
-    m_flx_subscription_store = sync::SubscriptionStore::create(m_db, [this](int64_t new_version) {
-        util::CheckedLockGuard lk(m_state_mutex);
-        if (m_state != State::Active && m_state != State::WaitingForAccessToken) {
-            return;
-        }
-        // There may be no session yet (i.e., waiting to refresh the access token).
-        if (m_session) {
-            m_session->on_new_flx_sync_subscription(new_version);
-        }
-    });
+
+    // Create the main subscription store instance when this is first called - this will
+    // remain valid afterwards for the life of the SyncSession, but m_flx_subscription_store
+    // will be reset when rolling back to PBS after a client FLX migration
+    if (!m_subscription_store_base) {
+        m_subscription_store_base = sync::SubscriptionStore::create(m_db, [this](int64_t new_version) {
+            util::CheckedLockGuard lk(m_state_mutex);
+            if (m_state != State::Active && m_state != State::WaitingForAccessToken) {
+                return;
+            }
+            // There may be no session yet (i.e., waiting to refresh the access token).
+            if (m_session) {
+                m_session->on_new_flx_sync_subscription(new_version);
+            }
+        });
+    }
+
+    // m_subscription_store_base is always around for the life of SyncSession, but the
+    // m_flx_subscription_store is set when using FLX.
+    m_flx_subscription_store = m_subscription_store_base;
```

Comment:
Do we even need to distinguish between this and m_flx_subscription_store and m_subscription_store_base?

Reply:
I was trying to find the simplest solution that wouldn't result in too many changes while keeping the subscription store around for PBS once it migrates to FLX and reverts back to PBS. Also, the subscription store is protected by the state mutex and m_config is protected by the config mutex, so both mutexes would need to be held at some point (maybe not at the same time) to read the flx requested state and return the subscription store, which could potentially cause deadlocks...
This solution retains the current usage of `m_flx_subscription_store`, while still retaining a cached shared ptr instance of the subscription store when rolling back to PBS.
Also, I was originally always creating the subscription store when a SyncSession was created, which causes a lockup if a write transaction is active before creating the SyncSession (like in one of our tests in _realm.cpp_), since the SubscriptionStore initializes the metadata version information if it doesn't exist when it is created.

Reply:
Okay, this is fine for now then. I wish wish wish that we could make the config immutable and get rid of that mutex.